src/library_exceptions.js

```diff
@@ -102,9 +102,17 @@ var LibraryExceptions = {
     }
   },
 
+  __cxa_init_primary_exception__deps: ['$ExceptionInfo'],
+  __cxa_init_primary_exception: (ptr, type, destructor) => {
+#if EXCEPTION_DEBUG
```

Comment:
This function looks like it could be a native symbol?

Reply:
Right now that this is not being used in `__cxa_throw` I think we can use the native version of this. But if I delete this `test_dylink_exceptions_try_catch_6` errors out with
```console
error: undefined symbol: __cxa_init_primary_exception (referenced by root reference (e.g. compiled C/C++ code))
warning: To disable errors for undefined symbols use `-sERROR_ON_UNDEFINED_SYMBOLS=0`
warning: ___cxa_init_primary_exception may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
```
I guess I need to add this symbol to some list of exported functions...? Can you help me where?

Reply:
I think if you add `__cxa_init_primary_exception` to `cxa_exception_emscripten.cpp` that should do the trick.  No need to export anything I would hope.

Reply:
Looks like this was added in https://github.com/llvm/llvm-project/pull/65534

Reply:
Done: https://github.com/emscripten-core/emscripten/pull/21638/commits/a4ec944dc5cd15b3c5263355189ecca496f5650a

Reply:
By removing this part of the code, I cannot compile previously working code due to the `error: undefined symbol: __cxa_init_primary_exception` when upgrading to emsdk 3.1.60

Reply:
It's quite a complex setup. It happens in [ITK library](https://github.com/InsightSoftwareConsortium/ITK).

You can find the whole reproducible setup in attached zip, which contains `Dockerfile` and a patch (few Emscripten fixes and external dependencies).
Just unzip and run `docker build .`

Error log:
````
208.4 [1598/1598] Linking CXX executable bin/itkTestDriver.js
208.4 FAILED: bin/itkTestDriver.js
208.4 : && /emsdk/upstream/emscripten/em++ -pthread -sSHARED_MEMORY=1  -Wall -Wcast-align -Wdisabled-optimization -Wextra -Wformat=2 -Winvalid-pch -Wno-format-nonliteral -Wpointer-arith -Wshadow -Wunused -Wwrite-strings -Wno-strict-overflow -Wno-deprecated -Wno-invalid-offsetof -Wno-undefined-var-template -Woverloaded-virtual  -O3 -DNDEBUG -pthread -sSHARED_MEMORY=1 -sMALLOC=mimalloc Modules/Core/TestKernel/src/CMakeFiles/itkTestDriver.dir/itkTestDriver.cxx.o -o bin/itkTestDriver.js  lib/libITKFFT-5.3.a  lib/libITKIOBMP-5.3.a  lib/libITKIOGDCM-5.3.a  lib/libITKIOGIPL-5.3.a  lib/libITKIOJPEG-5.3.a  lib/libITKIOMeshBYU-5.3.a  lib/libITKIOMeshFreeSurfer-5.3.a  lib/libITKIOMeshGifti-5.3.a  lib/libITKIOMeshOBJ-5.3.a  lib/libITKIOMeshOFF-5.3.a  lib/libITKIOMeshVTK-5.3.a  lib/libITKIOMeta-5.3.a  lib/libITKIONIFTI-5.3.a  lib/libITKIONRRD-5.3.a  lib/libITKIOPNG-5.3.a  lib/libITKIOTIFF-5.3.a  lib/libITKIOVTK-5.3.a  lib/libITKCommon-5.3.a  lib/libITKIOImageBase-5.3.a  lib/libITKTestKernel-5.3.a  lib/libITKFFT-5.3.a  lib/libITKIOBMP-5.3.a  lib/libITKIOGDCM-5.3.a  lib/libitkgdcmMSFF-5.3.a  lib/libitkgdcmDICT-5.3.a  lib/libitkgdcmIOD-5.3.a  lib/libitkgdcmDSED-5.3.a  lib/libitkgdcmCommon-5.3.a  lib/libitkgdcmjpeg8-5.3.a  lib/libitkgdcmjpeg12-5.3.a  lib/libitkgdcmjpeg16-5.3.a  lib/libitkgdcmopenjp2-5.3.a  lib/libitkgdcmcharls-5.3.a  lib/libitkgdcmuuid-5.3.a  lib/libITKIOGIPL-5.3.a  lib/libITKIOJPEG-5.3.a  lib/libITKIOMeshBYU-5.3.a  lib/libITKIOMeshFreeSurfer-5.3.a  lib/libITKIOMeshGifti-5.3.a  lib/libITKgiftiio-5.3.a  lib/libITKEXPAT-5.3.a  lib/libITKIOMeshOBJ-5.3.a  lib/libITKIOMeshOFF-5.3.a  lib/libITKIOMeshVTK-5.3.a  lib/libITKIOMeshBase-5.3.a  lib/libITKQuadEdgeMesh-5.3.a  lib/libITKMesh-5.3.a  lib/libITKIOMeta-5.3.a  lib/libITKMetaIO-5.3.a  lib/libITKIONIFTI-5.3.a  lib/libITKniftiio-5.3.a  lib/libITKznz-5.3.a  lib/libITKTransform-5.3.a  lib/libITKIONRRD-5.3.a  lib/libITKNrrdIO-5.3.a  lib/libITKIOPNG-5.3.a  lib/libitkpng-5.3.a  lib/libITKIOTIFF-5.3.a  lib/libitktiff-5.3.a  /sources/install/zlib/lib/libz.a  lib/libitkjpeg-5.3.a  lib/libITKIOVTK-5.3.a  lib/libITKIOImageBase-5.3.a  lib/libITKCommon-5.3.a  lib/libitkdouble-conversion-5.3.a  lib/libitksys-5.3.a  lib/libITKVNLInstantiation-5.3.a  lib/libitkvnl_algo-5.3.a  lib/libitkvnl-5.3.a  lib/libitkv3p_netlib-5.3.a  lib/libitkvcl-5.3.a  -lm  -lm && :
208.4 cache:INFO: generating system asset: symbol_lists/54eebcbad6cabbe30f5212b4867e5d79c17089ae.json... (this will be cached in "/emsdk/upstream/emscripten/cache/symbol_lists/54eebcbad6cabbe30f5212b4867e5d79c17089ae.json" for subsequent builds)
208.4 cache:INFO:  - ok
208.4 wasm-ld: error: lib/libITKCommon-5.3.a(itkPoolMultiThreader.cxx.o): undefined symbol: __cxa_init_primary_exception
208.4 em++: error: '/emsdk/upstream/bin/wasm-ld -o bin/itkTestDriver.wasm Modules/Core/TestKernel/src/CMakeFiles/itkTestDriver.dir/itkTestDriver.cxx.o lib/libITKFFT-5.3.a lib/libITKIOBMP-5.3.a lib/libITKIOGDCM-5.3.a lib/libITKIOGIPL-5.3.a lib/libITKIOJPEG-5.3.a lib/libITKIOMeshBYU-5.3.a lib/libITKIOMeshFreeSurfer-5.3.a lib/libITKIOMeshGifti-5.3.a lib/libITKIOMeshOBJ-5.3.a lib/libITKIOMeshOFF-5.3.a lib/libITKIOMeshVTK-5.3.a lib/libITKIOMeta-5.3.a lib/libITKIONIFTI-5.3.a lib/libITKIONRRD-5.3.a lib/libITKIOPNG-5.3.a lib/libITKIOTIFF-5.3.a lib/libITKIOVTK-5.3.a lib/libITKCommon-5.3.a lib/libITKIOImageBase-5.3.a lib/libITKTestKernel-5.3.a lib/libITKFFT-5.3.a lib/libITKIOBMP-5.3.a lib/libITKIOGDCM-5.3.a lib/libitkgdcmMSFF-5.3.a lib/libitkgdcmDICT-5.3.a lib/libitkgdcmIOD-5.3.a lib/libitkgdcmDSED-5.3.a lib/libitkgdcmCommon-5.3.a lib/libitkgdcmjpeg8-5.3.a lib/libitkgdcmjpeg12-5.3.a lib/libitkgdcmjpeg16-5.3.a lib/libitkgdcmopenjp2-5.3.a lib/libitkgdcmcharls-5.3.a lib/libitkgdcmuuid-5.3.a lib/libITKIOGIPL-5.3.a lib/libITKIOJPEG-5.3.a lib/libITKIOMeshBYU-5.3.a lib/libITKIOMeshFreeSurfer-5.3.a lib/libITKIOMeshGifti-5.3.a lib/libITKgiftiio-5.3.a lib/libITKEXPAT-5.3.a lib/libITKIOMeshOBJ-5.3.a lib/libITKIOMeshOFF-5.3.a lib/libITKIOMeshVTK-5.3.a lib/libITKIOMeshBase-5.3.a lib/libITKQuadEdgeMesh-5.3.a lib/libITKMesh-5.3.a lib/libITKIOMeta-5.3.a lib/libITKMetaIO-5.3.a lib/libITKIONIFTI-5.3.a lib/libITKniftiio-5.3.a lib/libITKznz-5.3.a lib/libITKTransform-5.3.a lib/libITKIONRRD-5.3.a lib/libITKNrrdIO-5.3.a lib/libITKIOPNG-5.3.a lib/libitkpng-5.3.a lib/libITKIOTIFF-5.3.a lib/libitktiff-5.3.a /sources/install/zlib/lib/libz.a lib/libitkjpeg-5.3.a lib/libITKIOVTK-5.3.a lib/libITKIOImageBase-5.3.a lib/libITKCommon-5.3.a lib/libitkdouble-conversion-5.3.a lib/libitksys-5.3.a lib/libITKVNLInstantiation-5.3.a lib/libitkvnl_algo-5.3.a lib/libitkvnl-5.3.a lib/libitkv3p_netlib-5.3.a lib/libitkvcl-5.3.a -L/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten /emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/crtbegin.o -lGL-mt-getprocaddr -lal -lhtml5 -lbulkmemory -lstubs -lnoexit -lc-mt -lmimalloc-mt -lcompiler_rt-mt -lc++-mt-noexcept -lc++abi-mt-noexcept -lsockets-mt -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr /tmp/tmpz32l3xi6libemscripten_js_symbols.so --import-memory --shared-memory --strip-debug --export=_emscripten_stack_alloc --export=_emscripten_thread_free_data --export=_emscripten_thread_crashed --export=emscripten_main_runtime_thread_id --export=emscripten_main_thread_process_queued_calls --export=_emscripten_run_on_main_thread_js --export=emscripten_stack_set_limits --export=__get_temp_ret --export=__set_temp_ret --export=__wasm_call_ctors --export=_emscripten_tls_init --export=_emscripten_thread_init --export=emscripten_stack_get_current --export=_emscripten_stack_restore --export=_emscripten_thread_exit --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__start_em_lib_deps --export-if-defined=__stop_em_lib_deps --export-if-defined=__start_em_js --export-if-defined=__stop_em_js --export-if-defined=main --export-if-defined=__main_argc_argv --export-table -z stack-size=65536 --no-growable-memory --initial-memory=16777216 --no-entry --table-base=1 --global-base=1024' failed (returned 1)
208.4 ninja: build stopped: subcommand failed.
```

Reply:
So the file that defines `__cxa_init_primary_exception` will only be included at link time if either wasm exceptions or emscripten exceptions are enabled.

See https://github.com/emscripten-core/emscripten/blob/27b30507ffd4141b57a6bf564280b78960ff617b/tools/system_libs.py#L1599-L1610

Is you can see from the link commend that is failing your link command is currently using `-lc++abi-mt-noexcept` which include `cxa_noexception.cpp`, which does not define this function.

I suspect that what is happening here is that `itkPoolMultiThreader.cxx.o` is being built with exception handling support but that the link command not including exception handling (e.g. via `-sDISABLE_EXCEPTION_CATCHING=0` or `-fexceptions`)?   Does this sounds correct @aheejin ?

Reply:
@mzukovec, are you able to find the compile command that was used to build `itkPoolMultiThreader.cxx.o`?

Reply:
@aheejin do you why some functions like `__cxa_rethrow_primary_exception` have stubs in `cxa_noexception.cpp`  but not `__cxa_init_primary_exception`

Reply:
@sbc100  from the `compile_commands.json`:
```
{
  "directory": "/sources/ITK-5.3.0/build",
  "command": "/emsdk/upstream/emscripten/em++ -I/sources/ITK-5.3.0/Modules/ThirdParty/DoubleConversion/src -I/sources/ITK-5.3.0/build/Modules/ThirdParty/DoubleConversion/src/double-conversion -I/sources/ITK-5.3.0/build/Modules/ThirdParty/Eigen3/src -I/sources/ITK-5.3.0/build/Modules/ThirdParty/KWSys/src -I/sources/ITK-5.3.0/Modules/ThirdParty/VNL/src/vxl/v3p/netlib -I/sources/ITK-5.3.0/Modules/ThirdParty/VNL/src/vxl/vcl -I/sources/ITK-5.3.0/Modules/ThirdParty/VNL/src/vxl/core -I/sources/ITK-5.3.0/build/Modules/ThirdParty/VNL/src/vxl/v3p/netlib -I/sources/ITK-5.3.0/build/Modules/ThirdParty/VNL/src/vxl/vcl -I/sources/ITK-5.3.0/build/Modules/ThirdParty/VNL/src/vxl/core -I/sources/ITK-5.3.0/build/Modules/Core/Common -I/sources/ITK-5.3.0/Modules/Core/Common/include -I/sources/ITK-5.3.0/Modules/ThirdParty/VNL/src/vxl/core/vnl/algo -I/sources/ITK-5.3.0/Modules/ThirdParty/VNL/src/vxl/core/vnl -I/sources/ITK-5.3.0/build/Modules/ThirdParty/VNL/src/vxl/core/vnl -isystem /sources/install/Eigen3/include/eigen3 -pthread -sSHARED_MEMORY=1  -Wall -Wcast-align -Wdisabled-optimization -Wextra -Wformat=2 -Winvalid-pch -Wno-format-nonliteral -Wpointer-arith -Wshadow -Wunused -Wwrite-strings -Wno-strict-overflow -Wno-deprecated -Wno-invalid-offsetof -Wno-undefined-var-template -Woverloaded-virtual  -O3 -DNDEBUG -fPIC -std=c++14 -o Modules/Core/Common/src/CMakeFiles/ITKCommon.dir/itkPoolMultiThreader.cxx.o -c /sources/ITK-5.3.0/Modules/Core/Common/src/itkPoolMultiThreader.cxx",
  "file": "/sources/ITK-5.3.0/Modules/Core/Common/src/itkPoolMultiThreader.cxx"
}
```
[compile_commands.json](https://github.com/emscripten-core/emscripten/files/15488929/compile_commands.json)

Reply:
@sbc100 
> I suspect that what is happening here is that `itkPoolMultiThreader.cxx.o` is being built with exception handling support but that the link command not including exception handling (e.g. via `-sDISABLE_EXCEPTION_CATCHING=0` or `-fexceptions`)? Does this sounds correct @aheejin ?

I thought so too, but the compile command @mzukovec gave didn't have any EH options, strange.

> @aheejin do you why some functions like `__cxa_rethrow_primary_exception` have stubs in `cxa_noexception.cpp` but not `__cxa_init_primary_exception`

I guess they are the only function called from libc++ where the code is not guarded with `#ifndef _LIBCPP_HAS_NO_EXCEPTIONS`. Apparently all uses of `__cxa_init_primary_exceptions` are guarded with `#ifndef _LIBCPP_HAS_NO_EXCEPTIONS`, so not sure what's causing the issue.

Reply:
@mzukovec Can you possibly share the `itkPoolMultiThreader.cxx.o` and `itkPoolMultiThreader.cxx`?

Reply:
> @mzukovec Can you possibly share the `itkPoolMultiThreader.cxx.o` and `itkPoolMultiThreader.cxx`?

Sure thing.
[itkPoolMultiThreader.zip](https://github.com/emscripten-core/emscripten/files/15495347/itkPoolMultiThreader.zip)


Link to the [used ITK revision](https://github.com/InsightSoftwareConsortium/ITK/tree/v5.3.0) in case it's needed later on.

Reply:
I think #22027 will fix this. Thanks for the report!